tools/chpldef/Message.cpp

```diff
@@ -331,86 +408,138 @@ const char* Message::errorToString(Error error) {
+    } else {
+      ctx_->message("The %s '%s' is complete...\n", dsc_, fmt_.c_str());
+    }
   }
 
-  ctx->message("Request '%s' complete...\n", msg->idToString().c_str());
+  template <Message::Behavior B>
+  void issueComputationCall() {}
+
+  template <>
+  void issueComputationCall<Message::INCOMING_REQUEST>() {
+    logComputationPrelude();
+    auto cr = M::compute(ctx_, msg_->p);
+    logComputationEpilogue(cr);
+    if (cr.isProgressingCallAgain) CHPLDEF_TODO();
+    if (cr.error != Message::OK) {
+      msg_->markFailed(cr.error, std::move(cr.note));
+    } else {
+      msg_->r = std::move(cr.result);
+      msg_->markCompleted();
+    }
+  }
 
-  return true;
-}
+  template <>
+  void issueComputationCall<Message::INCOMING_NOTIFY>() {
+    issueComputationCall<Message::INCOMING_REQUEST>();
+  }
 
-template <typename M, typename P, typename R>
-static bool
-doHandleMessage(Server* ctx, M* msg, const P& p, R& r) {
-  if (msg->isNotification()) return doHandleNotification(ctx, msg, p, r);
-  return doHandleRequest(ctx, msg, p, r);
-}
+  template <Message::Behavior B>
+  void issueComputationCall(Result r) {}
+
+  template <>
+  void issueComputationCall<Message::OUTBOUND_REQUEST>(Result r) {
+    logComputationPrelude();
+    msg_->r = std::move(r);
+    auto cr = M::compute(ctx_, msg_->r);
+    logComputationEpilogue(cr);
+    if (cr.isProgressingCallAgain) CHPLDEF_TODO();
+    if (cr.error != Message::OK) {
+      msg_->markFailed(cr.error, std::move(cr.note));
+    } else {
+      msg_->markCompleted();
+    }
+  }
 
-/** Call the request handler above and then mark completed or failed. */
-#define CHPLDEF_MESSAGE(name__, x1__, x2__, x3__) \
-  void name__::handle(Server* ctx) { \
-    ComputedResult r; \
-    if (doHandleMessage(ctx, this, p, r)) { \
-      this->r = std::move(r.result); \
-      this->markCompleted(); \
-    } else { \
-      this->markFailed(r.error, std::move(r.note)); \
-    } \
+  void handle() {
+    if (msg_->isOutbound() || msg_->isDone()) return;
+    if (msg_->status() == Message::PROGRESSING) CHPLDEF_TODO();
+    issueComputationCall<BEHAVIOR>();
   }
-#include "./message-macro-list.h"
-#undef CHPLDEF_MESSAGE
 
-opt<Response> Message::handle(Server* ctx, Message* msg) {
-  if (!ctx || !msg || msg->isResponse()) return {};
-  if (msg->status() != Message::PENDING) return {};
+  void handle(Response* rsp) {
+    CHPL_ASSERT(rsp && rsp->id() == msg_->id());
+    if (rsp->status() == Message::FAILED) CHPLDEF_TODO();
+    std::string note;
+    Result r;
+    auto error = M::unpack(rsp, r, &note);
+    if (error != Message::OK) {
+      msg_->markFailed(error, std::move(note));
+    } else {
+      handle(std::move(r));
+    }
+  }
 
-  msg->handle(ctx);
+  void handle(Result r) {
+    if (msg_->isOutbound() || msg_->isDone()) return;
+    if (msg_->status() == Message::PROGRESSING) CHPLDEF_TODO();
+    issueComputationCall<BEHAVIOR>(std::move(r));
+  }
+};
```

Comment:
There are just two variants, one which unpacks the Response JSON which is an operation that can fail, and another which takes a `Result` directly and just does the computation. I pass the `Response*` by pointer because that's the mode that these messages are usually being passed around.

Reply:
I've made a TODO to see if I can pass by reference instead.